nanshe_ipython.ipynb

```diff
@@ -545,6 +546,255 @@
+        "        )\n",
+        "\n",
+        "        # Persist frame shape\n",
+        "        frame_shape = frame_shape.persist()\n",
+        "\n",
+        "        # Compute the FFT of frames\n",
+        "        da_imgs_fft = fftn(da_imgs_flt, axes=tuple(irange(1, imgs.ndim)))\n",
+        "\n",
+        "        # Persist FFT of frames\n",
+        "        da_imgs_fft = da_imgs_fft.persist()\n",
+        "\n",
+        "        # Initialize\n",
+        "        i = 0\n",
+        "        avg_rel_dist = 1.0\n",
+        "        tmpl_hist_wght = da_imgs_flt.dtype.type(tmpl_hist_wght)\n",
+        "        shifts = da.zeros(\n",
+        "            (len(da_imgs_flt), da_imgs_flt.ndim - 1),\n",
+        "            dtype=int,\n",
+        "            chunks=(1, da_imgs_flt.ndim - 1)\n",
+        "        )\n",
+        "        da_imgs_fft_tmplt = da_imgs_fft.mean(axis=0)\n",
+        "\n",
+        "        # Persist FFT template image\n",
+        "        da_imgs_fft_tmplt = da_imgs_fft_tmplt.persist()\n",
+        "\n",
+        "        while avg_rel_dist > thld_rel_dist and i < num_reps:\n",
+        "            # Compute the shifted frames\n",
+        "            shifted_frames = []\n",
+        "            for j in irange(len(da_imgs_fft)):\n",
+        "                shifted_frames.append(dask_ndfourier.fourier_shift(\n",
+        "                    da_imgs_fft[i], shifts[i]\n",
+        "                ))\n",
+        "            shifted_frames = da.stack(shifted_frames)\n",
+        "\n",
+        "            # Compute the template FFT\n",
+        "            da_imgs_fft_tmplt = (\n",
+        "                tmpl_hist_wght * da_imgs_fft_tmplt +\n",
+        "                (1 - tmpl_hist_wght) * shifted_frames.mean(axis=0)\n",
+        "            )\n",
+        "\n",
+        "            # Free connected persisted values\n",
+        "            del shifted_frames\n",
+        "\n",
+        "            # Persist FFT template image\n",
+        "            da_imgs_fft_tmplt = da_imgs_fft_tmplt.persist()\n",
+        "\n",
+        "            # Find the best overlap with the template.\n",
+        "            overlap = ifftn(\n",
+        "                da_imgs_fft * da_imgs_fft_tmplt[None],\n",
+        "                axes=tuple(irange(1, imgs.ndim))\n",
+        "            ).real\n",
+        "            overlap_max = overlap.max(axis=tuple(irange(1, imgs.ndim)))\n",
+        "            overlap_max_match = (overlap == overlap_max[(Ellipsis,) + (None,) * (imgs.ndim - 1)])\n",
+        "\n",
+        "            # Compute the shift for each frame.\n",
+        "            old_shifts = shifts\n",
+        "            shifts = []\n",
+        "            for j in irange(len(overlap_max_match)):\n",
+        "                shift_j = dask.delayed(compute_offset)(overlap_max_match[j])\n",
+        "                shift_j = da.from_delayed(shift_j, (2,), int)\n",
```

Comment:
Appears the shift shape was hard-coded here. This was an oversight as the shift shape should be a function of the frame's dimensionality. Fixing the shift shape to be correctly determined in PR ( https://github.com/nanshe-org/nanshe_workflow/pull/132 ).